test/NuGet.Core.Tests/NuGet.ProjectManagement.Test/MSBuildNuGetProjectTests.cs

```diff
@@ -189,6 +189,121 @@ public class MSBuildNuGetProjectTests
+                var msBuildNuGetProjectSystem = new TestMSBuildNuGetProjectSystem(projectTargetFramework, testNuGetProjectContext);
+                var msBuildNuGetProject = new MSBuildNuGetProject(msBuildNuGetProjectSystem, randomPackagesFolderPath, randomPackagesConfigFolderPath);
+
+                // Pre-Assert
+                // Check that the packages.config file does not exist
+                Assert.False(File.Exists(randomPackagesConfigPath));
+                // Check that there are no packages returned by PackagesConfigProject
+                var packagesInPackagesConfig = (await msBuildNuGetProject.PackagesConfigNuGetProject.GetInstalledPackagesAsync(token)).ToList();
+                Assert.Equal(0, packagesInPackagesConfig.Count);
+                Assert.Equal(0, msBuildNuGetProjectSystem.References.Count);
+                Assert.Equal(0, msBuildNuGetProjectSystem.BatchCount);
+
+                var packageFileInfo = TestPackagesGroupedByFolder.GetLegacyTestPackageWithMultipleReferences(randomTestPackageSourcePath,
+                    packageIdentity.Id, packageIdentity.Version.ToNormalizedString());
+                using (var packageStream = GetDownloadResourceResult(packageFileInfo))
+                {
+                    // Act
+                    await msBuildNuGetProject.InstallPackageAsync(packageIdentity, packageStream, testNuGetProjectContext, token);
+                }
+
+                // Assert
+                // Check that the packages.config file exists after the installation
+                Assert.True(File.Exists(randomPackagesConfigPath));
+                // Check the number of packages and packages returned by PackagesConfigProject after the installation
+                packagesInPackagesConfig = (await msBuildNuGetProject.PackagesConfigNuGetProject.GetInstalledPackagesAsync(token)).ToList();
+                Assert.Equal(1, packagesInPackagesConfig.Count);
+                // Check that the reference has been added to MSBuildNuGetProjectSystem
+                Assert.Equal(2, msBuildNuGetProjectSystem.References.Count);
+                Assert.Equal(1, msBuildNuGetProjectSystem.BatchCount);
+                Assert.Equal("a.dll", msBuildNuGetProjectSystem.References.First().Key);
+                Assert.Equal("b.dll", msBuildNuGetProjectSystem.References.Skip(1).First().Key);
+                Assert.Equal(Path.Combine(msBuildNuGetProject.FolderNuGetProject.GetInstalledPath(packageIdentity),
+                    "lib\\net45\\a.dll"), msBuildNuGetProjectSystem.References.First().Value);
+                Assert.Equal(Path.Combine(msBuildNuGetProject.FolderNuGetProject.GetInstalledPath(packageIdentity),
+                    "lib\\net45\\b.dll"), msBuildNuGetProjectSystem.References.Skip(1).First().Value);
+            }
+        }
+
+        [Fact]
+        public async Task TestMSBuildNuGetProjectUninstallReferencesEventBatching()
+        {
```

Comment:
Please add test for case when install/ uninstall package fails after invoking batch start but still raise batch end event.